File: src/simulation/ai/llm/RecordReplayCache.js
// RecordReplayCache — VCR-cassette LLM cache for the academic benchmark.
//
// Modes:
//   - "record"  — every get() returns null; the caller is expected to run
//                 the real adapter and then call put(req, response). The
//                 cache writes (key, request_summary, response, timestamp)
//                 to an NDJSON cassette on flush().
//   - "replay"  — every get() looks the request up in the cassette and
//                 returns the cached response, OR throws
//                 RecordReplayCacheMiss on miss. put() is a no-op.
//   - "auto"    — replay first; on miss, behaves like record (returns
//                 null from get(), accepts put()). Useful for dev runs
//                 where you want to top up the cassette without nuking
//                 it.
//
// Cassette format: NDJSON, one JSON object per line:
//   { "key": "<sha256>", "request_summary": {...}, "response": {...},
//     "timestamp": "<ISO>" }
//
// Cache key canonicalization:
//   sha256(JSON.stringify({channel, model, prompt, temperature, top_p}))
// with keys serialized in a fixed order. This guarantees the same
// (channel,model,prompt,temp,top_p) tuple always hashes identically
// regardless of object property order.
//
// Adapter wrap helper: `wrapAdapterWithCache(adapter, cache)` returns an
// AgentAdapter that consults the cache before delegating, and records
// adapter responses to the cache on miss.

import { createHash } from "node:crypto";
import { promises as fs } from "node:fs";
import path from "node:path";

import { AgentAdapter } from "./AgentAdapter.js";
import {
  buildEnvironmentPromptUserContent,
  buildPolicyPromptUserContent,
} from "./PromptPayload.js";

export class RecordReplayCacheMiss extends Error {
  constructor(key, summary) {
    super(`record-replay miss: key=${key}`);
    this.name = "RecordReplayCacheMiss";
    this.key = key;
    this.summary = summary;
  }
}

const VALID_MODES = new Set(["record", "replay", "auto"]);

// Property order matters for sha256 stability — we hand-serialize.
function canonicalizeRequest(req) {
  return {
    channel: String(req?.channel ?? ""),
    model: String(req?.model ?? ""),
    prompt: typeof req?.prompt === "string" ? req.prompt : JSON.stringify(req?.prompt ?? null),
    temperature: Number.isFinite(Number(req?.temperature)) ? Number(req.temperature) : null,
    top_p: Number.isFinite(Number(req?.top_p)) ? Number(req.top_p) : null,
  };
}

function canonicalKeyJson(req) {
  const c = canonicalizeRequest(req);
  // Hand-built JSON to fix property order across engines.
  return `{"channel":${JSON.stringify(c.channel)},`
       + `"model":${JSON.stringify(c.model)},`
       + `"prompt":${JSON.stringify(c.prompt)},`
       + `"temperature":${JSON.stringify(c.temperature)},`
       + `"top_p":${JSON.stringify(c.top_p)}}`;
}

function summarizeRequest(req) {
  const c = canonicalizeRequest(req);
  return {
    channel: c.channel,
    model: c.model,
    promptLength: c.prompt.length,
    temperature: c.temperature,
    top_p: c.top_p,
  };
}

export class RecordReplayCache {
  /**
   * @param {string} cassettePath — file path for NDJSON cassette
   * @param {"record"|"replay"|"auto"} [mode="replay"]
   */
  constructor(cassettePath, mode = "replay") {
    if (!cassettePath || typeof cassettePath !== "string") {
      throw new Error("RecordReplayCache: cassettePath required");
    }
    if (!VALID_MODES.has(mode)) {
      throw new Error(`RecordReplayCache: invalid mode "${mode}"; expected one of ${[...VALID_MODES].join(", ")}`);
    }
    this.cassettePath = cassettePath;
    this.mode = mode;
    /** @type {Map<string, {request_summary:object, response:object, timestamp:string}>} */
    this._entries = new Map();
    this._loaded = false;
    this._dirty = false;
    this._stats = { hits: 0, misses: 0, total: 0 };
  }

  /** sha256 hex digest of canonicalized request. */
  computeKey(req) {
    const json = canonicalKeyJson(req);
    return createHash("sha256").update(json).digest("hex");
  }

  async _ensureLoaded() {
    if (this._loaded) return;
    this._loaded = true;
    try {
      const txt = await fs.readFile(this.cassettePath, "utf8");
      const lines = txt.split(/\r?\n/);
      for (const line of lines) {
        const trimmed = line.trim();
        if (!trimmed) continue;
        try {
          const entry = JSON.parse(trimmed);
          if (entry && typeof entry.key === "string" && entry.response !== undefined) {
            this._entries.set(entry.key, {
              request_summary: entry.request_summary ?? null,
              response: entry.response,
              timestamp: entry.timestamp ?? "",
            });
          }
        } catch {
          // skip malformed line
        }
      }
    } catch (err) {
      if (err && err.code === "ENOENT") {
        // Fresh cassette — no entries yet.
        return;
      }
      throw err;
    }
  }

  /**
   * @param {object} req
   * @returns {Promise<object|null>} cached DecisionResponse or null on miss
   */
  async get(req) {
    await this._ensureLoaded();
    const key = this.computeKey(req);
    this._stats.total += 1;
    const entry = this._entries.get(key);

    if (this.mode === "record") {
      // Record mode never returns cached entries — caller will record
      // after running the adapter.
      this._stats.misses += 1;
      return null;
    }

    if (entry) {
      this._stats.hits += 1;
      // Defensive copy so callers can mutate freely.
      return JSON.parse(JSON.stringify(entry.response));
    }

    this._stats.misses += 1;

    if (this.mode === "replay") {
      throw new RecordReplayCacheMiss(key, summarizeRequest(req));
    }
    // auto: miss falls through to record-on-put.
    return null;
  }

  /**
   * @param {object} req
   * @param {object} response
   */
  async put(req, response) {
    if (this.mode === "replay") {
      // Strict replay never records.
      return;
    }
    await this._ensureLoaded();
    const key = this.computeKey(req);
    this._entries.set(key, {
      request_summary: summarizeRequest(req),
      response: JSON.parse(JSON.stringify(response ?? null)),
      timestamp: new Date().toISOString(),
    });
    this._dirty = true;
  }

  async flush() {
    if (this.mode === "replay") return;
    if (!this._dirty && this._entries.size === 0) {
      // Still ensure the file exists so downstream replay can find it.
      try { await fs.mkdir(path.dirname(this.cassettePath), { recursive: true }); } catch { /* noop */ }
      try { await fs.writeFile(this.cassettePath, "", "utf8"); } catch { /* noop */ }
      return;
    }
    const lines = [];
    for (const [key, entry] of this._entries) {
      lines.push(JSON.stringify({
        key,
        request_summary: entry.request_summary,
        response: entry.response,
        timestamp: entry.timestamp,
      }));
    }
    try {
      await fs.mkdir(path.dirname(this.cassettePath), { recursive: true });
    } catch { /* noop */ }
    await fs.writeFile(this.cassettePath, lines.join("\n") + (lines.length ? "\n" : ""), "utf8");
    this._dirty = false;
  }

  stats() {
    const { hits, misses, total } = this._stats;
    return {
      hits,
      misses,
      total,
      hitRate: total > 0 ? hits / total : 0,
    };
  }
}

// ----- adapter wrap helper -----

function inferRequestKey(channel, payload, options) {
  const promptStr = (() => {
    if (channel === "environment-director") return buildEnvironmentPromptUserContent(payload);
    if (channel === "npc-policy") return buildPolicyPromptUserContent(payload);
    if (typeof payload === "string") return payload;
    try { return JSON.stringify(payload); } catch { return String(payload); }
  })();
  return {
    channel,
    model: String(options?.model ?? ""),
    prompt: promptStr,
    temperature: Number.isFinite(Number(options?.temperature)) ? Number(options.temperature) : null,
    top_p: Number.isFinite(Number(options?.top_p)) ? Number(options.top_p) : null,
  };
}

class CachedAdapter extends AgentAdapter {
  constructor(adapter, cache) {
    super();
    this._adapter = adapter;
    this._cache = cache;
  }

  async request(channel, payload, options = {}) {
    const cacheReq = inferRequestKey(channel, payload, options);
    let cached = null;
    try {
      cached = await this._cache.get(cacheReq);
    } catch (err) {
      if (err instanceof RecordReplayCacheMiss) {
        // Strict replay mode and miss — surface the miss.
        throw err;
      }
      throw err;
    }
    if (cached) return cached;

    const response = await this._adapter.request(channel, payload, options);
    try {
      await this._cache.put(cacheReq, response);
    } catch {
      // best-effort
    }
    return response;
  }
}

/**
 * @param {AgentAdapter} adapter
 * @param {RecordReplayCache} cache
 * @returns {AgentAdapter}
 */
export function wrapAdapterWithCache(adapter, cache) {
  if (!adapter || typeof adapter.request !== "function") {
    throw new Error("wrapAdapterWithCache: adapter.request required");
  }
  if (!(cache instanceof RecordReplayCache)) {
    throw new Error("wrapAdapterWithCache: cache must be a RecordReplayCache instance");
  }
  return new CachedAdapter(adapter, cache);
}

export default RecordReplayCache;

File: project_utopia/simulation/ai/llm/record_replay_cache.py
"""RecordReplayCache — VCR-cassette LLM cache for the academic benchmark.

Python port of ``RecordReplayCache.js``. Three modes:

- ``record`` — every ``get()`` returns ``None``; the caller runs the real
  adapter and then ``await cache.put(req, response)`` records it. ``flush()``
  writes the cassette as NDJSON.
- ``replay`` — every ``get()`` looks the request up in the cassette and
  returns the cached response, OR raises :class:`RecordReplayCacheMiss` on
  miss. ``put()`` is a no-op.
- ``auto`` — replay first; on miss, falls back to record-on-put. Useful for
  dev runs that want to top up the cassette without nuking it.

Key canonicalisation: ``sha256(JSON.stringify({channel, model, prompt,
temperature, top_p}))`` with hand-built JSON for stable key ordering across
languages.

The default on-disk layout is ``.cache/llm-cassettes/{hash}.json`` per the
Phase 1 design brief, but the constructor accepts an explicit cassette path
to remain a drop-in match for the JS module.
"""

from __future__ import annotations

import asyncio
import hashlib
import json
import os
from pathlib import Path
from typing import Any

from .agent_adapter import AgentAdapter, Channel, DecisionResponse
from .prompt_payload import (
    build_environment_prompt_user_content,
    build_policy_prompt_user_content,
)

VALID_MODES = ("record", "replay", "auto")


class RecordReplayCacheMiss(Exception):
    """Raised by :class:`RecordReplayCache.get` in ``replay`` mode on miss."""

    def __init__(self, key: str, summary: dict[str, Any]) -> None:
        super().__init__(f"record-replay miss: key={key}")
        self.key = key
        self.summary = summary


def _canonicalize_request(req: dict[str, Any]) -> dict[str, Any]:
    prompt = req.get("prompt")
    if not isinstance(prompt, str):
        try:
            prompt = json.dumps(prompt, sort_keys=True, separators=(",", ":"))
        except (TypeError, ValueError):
            prompt = str(prompt)
    temperature = req.get("temperature")
    top_p = req.get("top_p")
    try:
        temperature = float(temperature) if temperature is not None else None
    except (TypeError, ValueError):
        temperature = None
    try:
        top_p = float(top_p) if top_p is not None else None
    except (TypeError, ValueError):
        top_p = None
    return {
        "channel": str(req.get("channel") or ""),
        "model": str(req.get("model") or ""),
        "prompt": prompt,
        "temperature": temperature,
        "top_p": top_p,
    }


def _canonical_key_json(req: dict[str, Any]) -> str:
    c = _canonicalize_request(req)
    # Hand-built JSON to lock field order across languages.
    return (
        "{"
        f'"channel":{json.dumps(c["channel"])},'
        f'"model":{json.dumps(c["model"])},'
        f'"prompt":{json.dumps(c["prompt"])},'
        f'"temperature":{json.dumps(c["temperature"])},'
        f'"top_p":{json.dumps(c["top_p"])}'
        "}"
    )


def _summarize_request(req: dict[str, Any]) -> dict[str, Any]:
    c = _canonicalize_request(req)
    return {
        "channel": c["channel"],
        "model": c["model"],
        "promptLength": len(c["prompt"]),
        "temperature": c["temperature"],
        "top_p": c["top_p"],
    }


class RecordReplayCache:
    """Hash-keyed VCR cassette over an NDJSON file.

    Args:
        cassette_path: Path to the cassette file. Parent directories are
            created on flush.
        mode: One of ``"record"``, ``"replay"``, ``"auto"``.
    """

    def __init__(self, cassette_path: str | os.PathLike[str], mode: str = "replay") -> None:
        if not cassette_path:
            raise ValueError("RecordReplayCache: cassette_path required")
        if mode not in VALID_MODES:
            raise ValueError(
                f"RecordReplayCache: invalid mode {mode!r}; expected one of {VALID_MODES}"
            )
        self.cassette_path = Path(cassette_path)
        self.mode = mode
        self._entries: dict[str, dict[str, Any]] = {}
        self._loaded = False
        self._dirty = False
        self._stats = {"hits": 0, "misses": 0, "total": 0}
        self._lock = asyncio.Lock()

    def compute_key(self, req: dict[str, Any]) -> str:
        return hashlib.sha256(_canonical_key_json(req).encode("utf-8")).hexdigest()

    async def _ensure_loaded(self) -> None:
        if self._loaded:
            return
        self._loaded = True
        try:
            text = self.cassette_path.read_text(encoding="utf-8")
        except FileNotFoundError:
            return
        for line in text.splitlines():
            line = line.strip()
            if not line:
                continue
            try:
                entry = json.loads(line)
            except (ValueError, TypeError):
                continue
            if not isinstance(entry, dict):
                continue
            key = entry.get("key")
            if not isinstance(key, str) or "response" not in entry:
                continue
            self._entries[key] = {
                "request_summary": entry.get("request_summary"),
                "response": entry["response"],
                "timestamp": entry.get("timestamp", ""),
            }

    async def get(self, req: dict[str, Any]) -> dict[str, Any] | None:
        async with self._lock:
            await self._ensure_loaded()
            key = self.compute_key(req)
            self._stats["total"] += 1
            entry = self._entries.get(key)

            if self.mode == "record":
                self._stats["misses"] += 1
                return None
            if entry:
                self._stats["hits"] += 1
                return json.loads(json.dumps(entry["response"]))  # defensive copy
            self._stats["misses"] += 1
            if self.mode == "replay":
                raise RecordReplayCacheMiss(key, _summarize_request(req))
            return None

    async def put(self, req: dict[str, Any], response: Any) -> None:
        if self.mode == "replay":
            return
        async with self._lock:
            await self._ensure_loaded()
            key = self.compute_key(req)
            self._entries[key] = {
                "request_summary": _summarize_request(req),
                "response": json.loads(json.dumps(response if response is not None else None)),
                "timestamp": _iso_now(),
            }
            self._dirty = True

    async def flush(self) -> None:
        if self.mode == "replay":
            return
        async with self._lock:
            self.cassette_path.parent.mkdir(parents=True, exist_ok=True)
            if not self._dirty and not self._entries:
                # Touch an empty cassette so downstream replay can find it.
                self.cassette_path.write_text("", encoding="utf-8")
                return
            lines: list[str] = []
            for key, entry in self._entries.items():
                lines.append(
                    json.dumps(
                        {
                            "key": key,
                            "request_summary": entry["request_summary"],
                            "response": entry["response"],
                            "timestamp": entry["timestamp"],
                        }
                    )
                )
            payload = "\n".join(lines) + ("\n" if lines else "")
            self.cassette_path.write_text(payload, encoding="utf-8")
            self._dirty = False

    def stats(self) -> dict[str, Any]:
        hits = self._stats["hits"]
        misses = self._stats["misses"]
        total = self._stats["total"]
        return {
            "hits": hits,
            "misses": misses,
            "total": total,
            "hitRate": (hits / total) if total else 0.0,
        }


def _iso_now() -> str:
    from datetime import datetime, timezone
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.") + f"{datetime.now(timezone.utc).microsecond // 1000:03d}Z"


# ---------------------------------------------------------------------------
# Adapter wrapper
# ---------------------------------------------------------------------------


def _infer_request_key(channel: Channel, payload: Any, options: dict[str, Any] | None) -> dict[str, Any]:
    if channel == "environment-director":
        prompt_str = build_environment_prompt_user_content(payload or {})
    elif channel == "npc-policy":
        prompt_str = build_policy_prompt_user_content(payload or {})
    elif isinstance(payload, str):
        prompt_str = payload
    else:
        try:
            prompt_str = json.dumps(payload, sort_keys=True)
        except (TypeError, ValueError):
            prompt_str = str(payload)
    opts = options or {}
    return {
        "channel": channel,
        "model": str(opts.get("model") or ""),
        "prompt": prompt_str,
        "temperature": opts.get("temperature"),
        "top_p": opts.get("top_p"),
    }


class _CachedAdapter(AgentAdapter):
    def __init__(self, adapter: AgentAdapter, cache: RecordReplayCache) -> None:
        self._adapter = adapter
        self._cache = cache

    async def request(
        self,
        channel: Channel,
        payload: Any,
        options: dict[str, Any] | None = None,
    ) -> DecisionResponse:
        cache_req = _infer_request_key(channel, payload, options)
        cached: dict[str, Any] | None = None
        try:
            cached = await self._cache.get(cache_req)
        except RecordReplayCacheMiss as err:
            return DecisionResponse(
                data=None,
                fallback=True,
                latency_ms=0.0,
                model="record-replay-miss",
                error=f"record-replay miss: {err}",
                debug={"channel": channel, "cacheKey": self._cache.compute_key(cache_req)},
            )
        except Exception as err:  # noqa: BLE001
            return DecisionResponse(
                data=None,
                fallback=True,
                latency_ms=0.0,
                model="record-replay-error",
                error=f"record-replay error: {err}",
                debug={"channel": channel},
            )
        if cached is not None:
            try:
                return DecisionResponse.model_validate(cached)
            except Exception:  # noqa: BLE001
                return DecisionResponse(
                    data=cached.get("data") if isinstance(cached, dict) else None,
                    fallback=bool(cached.get("fallback")) if isinstance(cached, dict) else True,
                    latency_ms=float(cached.get("latencyMs") or 0.0) if isinstance(cached, dict) else 0.0,
                    model=str(cached.get("model") or "") if isinstance(cached, dict) else "",
                    error=str(cached.get("error") or "") if isinstance(cached, dict) else "",
                    debug=cached.get("debug") or {},
                )

        response = await self._adapter.request(channel, payload, options)
        try:
            await self._cache.put(cache_req, response.model_dump(by_alias=True))
        except Exception:  # noqa: BLE001 - best-effort record
            pass
        return response


def wrap_adapter_with_cache(adapter: AgentAdapter, cache: RecordReplayCache) -> AgentAdapter:
    """Return an :class:`AgentAdapter` that consults ``cache`` before delegating."""
    if not isinstance(adapter, AgentAdapter):
        raise TypeError("wrap_adapter_with_cache: adapter must subclass AgentAdapter")
    if not isinstance(cache, RecordReplayCache):
        raise TypeError("wrap_adapter_with_cache: cache must be a RecordReplayCache")
    return _CachedAdapter(adapter, cache)


__all__ = [
    "RecordReplayCache",
    "RecordReplayCacheMiss",
    "VALID_MODES",
    "wrap_adapter_with_cache",
]
